aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/ECSComputeServiceAdapter.java

```diff
@@ -175,33 +176,25 @@ public boolean apply(@Nullable InstanceType input) {
    }
 
    @Override
-   public Iterable<Image> listImages() {
-      final ImmutableList.Builder<Image> images = ImmutableList.builder();
-      final List<String> availableLocationNames = newArrayList(
-              transform(listLocations(), new Function<Region, String>() {
-                 @Override
-                 public String apply(Region location) {
-                    return location.id();
-                 }
-              }));
-
-      for (String regionId : availableLocationNames) {
-         images.addAll(api.imageApi().list(regionId).concat());
+   public Iterable<ImageInRegion> listImages() {
+      final ImmutableList.Builder<ImageInRegion> imagesInRegion = ImmutableList.builder();
+
+      for (final String regionId : getAvailableLocationNames()) {
+               imagesInRegion.addAll(Iterables.transform(api.imageApi().list(regionId).concat(), new Function<Image, ImageInRegion>() {
+                  @Override
+                  public ImageInRegion apply(Image image) {
+                     return ImageInRegion.create(regionId, image);
+                  }
+               }));
       }
-      return images.build();
+      return imagesInRegion.build();
    }
 
    @Override
-   public Image getImage(final String id) {
-      Optional<Image> firstInterestingImage = Iterables.tryFind(listImages(), new Predicate<Image>() {
-         public boolean apply(Image input) {
-            return input.id().equals(id);
-         }
-      });
-      if (!firstInterestingImage.isPresent()) {
-         throw new IllegalStateException("Cannot find image with the required slug " + id);
-      }
-      return firstInterestingImage.get();
+   public ImageInRegion getImage(final String id) {
+      RegionAndId regionAndId = fromSlashEncoded(id);
+      return ImageInRegion.create(regionAndId.regionId(), Iterables.getFirst(api.imageApi().list(regionAndId.regionId(),
```

Comment:
This returns an object even if the image does not exist. That-s not correct. Just return `null`.